### PR TITLE
UT: import gomock to make write ut easier

### PR DESCRIPTION
**Pull request by jwcesign**
Signed-off-by: jwcesign <jiangwei115@huawei.com>

**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:
Give a demo how to use gomock to make write ut easier

**Which issue(s) this PR fixes**:
Part of https://github.com/karmada-io/karmada/issues/2579

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
none
```

### CODE REVIEW

File: hack/update-mocks.sh [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env bash
+
```
Comment (chaunceyjiang) [inline hack/update-mocks.sh:2]: image

Comment (chaunceyjiang) [reply hack/update-mocks.sh:2]: In my local environment.

File: hack/update-mocks.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        -o -wholename '*/bindata.go' \
+      \) -prune \
+    \) -name '*.go'
+}
+
+cd "${KARMADA_ROOT}"
+
+echo 'executing go generate command on below files'
+for IFILE in $(find_files | xargs grep --files-with-matches -e '//go:generate mockgen'); do
+    go generate -v "$IFILE"
```
Comment (chaunceyjiang) [inline hack/update-mocks.sh:51]: image


Can we check the `mockgen` command in this script?

File: hack/update-mocks.sh (hunk 1 above)
Comment (jwcesign) [reply hack/update-mocks.sh:2]: Ops, I will chmod it

File: hack/update-mocks.sh (hunk 2 above)
Comment (jwcesign) [reply hack/update-mocks.sh:51]: Sure, it's better idea.

Comment (jwcesign) [reply hack/update-mocks.sh:56]: Fix, try again

**Comment by chaunceyjiang** [comment]
lint failed.

https://github.com/karmada-io/karmada/actions/runs/3287561163/jobs/5416867607

**Comment by jwcesign** [comment]
> lint failed.
> 
> https://github.com/karmada-io/karmada/actions/runs/3287561163/jobs/5416867607

If this idea is received, should change link check to ignore mock gen  files.

File: hack/update-mocks.sh [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+KARMADA_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+# Explicitly opt into go modules, even though we're inside a GOPATH directory
+export GO111MODULE=on
+export PATH=$PATH:$(go env GOPATH)/bin
+
+trap EXIT
+
+echo 'installing mockgen'
+go install github.com/golang/mock/mockgen
```
Comment (chaunceyjiang) [inline hack/update-mocks.sh:32]:
```suggestion
source "${KARMADA_ROOT}"/hack/util.sh
echo -n "Preparing: 'mockgen' existence check - "
if util::cmd_exist mockgen; then
  echo "passed"
else
  echo "not pass"
  util::install_tools "github.com/golang/mock/mockgen" "v1.6.0"
fi
```

**Comment by chaunceyjiang** [comment]
> If this idea is received, should change link check to ignore mock gen files.

Get it.

File: pkg/scheduler/framework/runtime/framework_test.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+			result := frameWork.RunFilterPlugins(ctx, nil, nil, nil)
+			if result.IsSuccess() != tt.isSuccess {
+				t.Errorf("want %v, but get:%v", tt.isSuccess, result.IsSuccess())
+			}
+		})
+	}
+}
+
+func createAndRegisterFactory(plugin framework.Plugin) (Registry, error) {
+	filterPluginFactory := PluginFactory(func() (framework.Plugin, error) {
```
Comment (RainbowMango) [inline pkg/scheduler/framework/runtime/framework_test.go:66]: Redundant type conversion

Comment (RainbowMango) [reply pkg/scheduler/framework/runtime/framework_test.go:66]: Maybe we don't need to abstract the function.

**Comment by jwcesign** [comment]
cc @RainbowMango PTAL

**Comment by jwcesign** [comment]
> > If this idea is received, should change link check to ignore mock gen files.
> 
> Get it.

I change the link check tools, please take a loook, @chaunceyjiang

File: hack/tools/preferredimports/preferredimports.go [hunk 5]
```diff
@@ -42,12 +42,13 @@ import (
...
-	importAliases = flag.String("import-aliases", "hack/.import-aliases", "json file with import aliases")
-	confirm       = flag.Bool("confirm", false, "update file with the preferred aliases for imports")
-	regex         = flag.String("include-path", "(test/e2e/|test/e2e_node)", "only files with paths matching this regex is touched")
-	isTerminal    = term.IsTerminal(int(os.Stdout.Fd()))
-	logPrefix     = ""
-	aliases       map[string]string
+	importAliases    = flag.String("import-aliases", "hack/.import-aliases", "json file with import aliases")
+	confirm          = flag.Bool("confirm", false, "update file with the preferred aliases for imports")
+	includePathRegex = flag.String("include-path", "(test/e2e/|test/e2e_node)", "only files with paths matching this regex is touched")
+	excludePathRegex = flag.String("exclude-path", "(testing)", "exclude path for check alias")
```
Comment (RainbowMango) [inline hack/tools/preferredimports/preferredimports.go:48]:
```suggestion
	excludePathRegex = flag.String("exclude-path", "(testing)", "files with paths matching this regex is ignored")
```

File: hack/update-mocks.sh [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+export GO111MODULE=on
+export PATH=$PATH:$(go env GOPATH)/bin
+
+trap EXIT
+
+echo 'installing mockgen'
+source "${KARMADA_ROOT}"/hack/util.sh
+echo -n "Preparing: 'mockgen' existence check - "
+if [ ! $(util::cmd_exist mockgen) ]; then
+  util::install_tools "github.com/golang/mock/mockgen" "latest"
```
Comment (RainbowMango) [inline hack/update-mocks.sh:35]: We can't install `latest`, the generated file may change over `mockgen` versions.
```suggestion
  GO111MODULE=on go install "github.com/golang/mock/mockgen"
```
This way installing `mockgen` with the version in the vendor. For CI we need a consistent version.

**Review by RainbowMango — COMMENTED** [review]
Generally looks good to me.

File: hack/update-mocks.sh (hunk 6 above)
Comment (chaunceyjiang) [reply hack/update-mocks.sh:35]: +1

File: pkg/scheduler/framework/runtime/framework_test.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			frameWork, err := NewFramework(registry)
+			if err != nil {
+				t.Errorf("create frame work error:%v", err)
+			}
+
+			tt.mockFunc()
```
Comment (cmicat) [inline pkg/scheduler/framework/runtime/framework_test.go:140]: I think every test case should be independent. So it's better not share mock object between test cases. Instead, we can create a new mock object in every test case.

File: pkg/scheduler/framework/runtime/framework_test.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+			},
+			isSuccess: true,
+		},
+		{
+			name: "Test filter func error",
+			mockFunc: func() {
+				mockFilterPlugin.EXPECT().Filter(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+					Return(framework.NewResult(framework.Error, "foo error"))
+			},
+			isSuccess: false,
```
Comment (ikaven1024) [inline pkg/scheduler/framework/runtime/framework_test.go:40]: It only test two cases: with one error or error plugin. Other cases are not covered: no plugin, mix plugins.

How about these:
```go
func Test_frameworkImpl_RunFilterPlugins(t *testing.T) {
	ctx := context.Background()
	mockCtrl := gomock.NewController(t)
	defer mockCtrl.Finish()

	alwaysError := frameworktesting.NewMockFilterPlugin(mockCtrl)
	alwaysError.EXPECT().Filter(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(framework.NewResult(framework.Error))

	alwaysSuccess := frameworktesting.NewMockFilterPlugin(mockCtrl)
	alwaysSuccess.EXPECT().Filter(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(framework.NewResult(framework.Success))

	tests := []struct {
		name      string
		framework framework.Framework
		isSuccess bool
	}{
		{
			name: "no plugin",
			framework: &frameworkImpl{
				filterPlugins: []framework.FilterPlugin{},
			},
			isSuccess: true,
		},
		{
			name: "error plugin",
			framework: &frameworkImpl{
				filterPlugins: []framework.FilterPlugin{alwaysError},
			},
			isSuccess: false,
		},
		{
			name: "success plugin",
			framework: &frameworkImpl{
				filterPlugins: []framework.FilterPlugin{alwaysSuccess},
			},
			isSuccess: true,
		},
		{
			name: "error and success plugins",
			framework: &frameworkImpl{
				filterPlugins: []framework.FilterPlugin{alwaysError, alwaysSuccess},
			},
			isSuccess: false,
		},
		{
			name: "success and error plugins",
			framework: &frameworkImpl{
				filterPlugins: []framework.FilterPlugin{alwaysSuccess, alwaysError},
			},
			isSuccess: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := tt.framework.RunFilterPlugins(ctx, nil, nil, nil)
			if result.IsSuccess() != tt.isSuccess {
				t.Errorf("want %v, but get:%v", tt.isSuccess, result.IsSuccess())
			}
		})
	}
}
```

File: hack/update-mocks.sh [hunk 9]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env bash
```
Comment (ikaven1024) [inline hack/update-mocks.sh:1]: please add `verify-mocks.sh`, add add these files to `update-all.sh` and `verify-all.sh`.

File: pkg/scheduler/framework/runtime/framework_test.go [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+	}
+}
+
+func createAndRegisterFactory(plugins []framework.Plugin) (Registry, error) {
+	registry := Registry{}
+	for _, plugin := range plugins {
+		filterPluginFactory := func() (framework.Plugin, error) {
+			return plugin, nil
+		}
+		if err := registry.Register("foo", filterPluginFactory); err != nil {
```
Comment (ikaven1024) [inline pkg/scheduler/framework/runtime/framework_test.go:84]: we can not register plugins with the same name, they are override.

File: hack/verify-all.sh [hunk 11]
```diff
@@ -19,3 +19,4 @@ bash "$REPO_ROOT/hack/verify-vendor.sh"
 bash "$REPO_ROOT/hack/verify-swagger-docs.sh"
 bash "$REPO_ROOT/hack/verify-crdgen.sh"
 bash "$REPO_ROOT/hack/verify-codegen.sh"
+bash "$REPO_ROOT/hack/verify-mocks.sh"
```
Comment (ikaven1024) [inline hack/verify-all.sh:22]: place the script in order with:
https://github.com/karmada-io/karmada/blob/5a1addf6936d7c6604fb26c4ae725ce7c716db3a/hack/verify-all.sh#L12-L14

File: pkg/scheduler/framework/runtime/framework_test.go (hunk 8 above)
Comment (jwcesign) [reply pkg/scheduler/framework/runtime/framework_test.go:39]: done for filter plugin, but not for score plugin, it does not increase code coverage and get few benefits.

File: pkg/scheduler/framework/runtime/framework_test.go (hunk 10 above)
Comment (jwcesign) [reply pkg/scheduler/framework/runtime/framework_test.go:84]: Every time createAndRegisterFactory is called, the registry is created new, so there is no override.

File: pkg/scheduler/framework/runtime/framework_test.go (hunk 7 above)
Comment (jwcesign) [reply pkg/scheduler/framework/runtime/framework_test.go:155]: It make sense, but I prefer to keep it as case global, it makes code looks more simple

**Comment by jwcesign** [comment]
cc @RainbowMango @ikaven1024 @chaunceyjiang @cmicat PTAL

File: pkg/scheduler/framework/runtime/framework_test.go (hunk 7 above)
Comment (ikaven1024) [reply pkg/scheduler/framework/runtime/framework_test.go:163]: mock is polluted between cases, It doesn't seem a good thing.

File: pkg/scheduler/framework/runtime/framework_test.go [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+	for _, plugin := range plugins {
+		pluginCopy := plugin
+		filterPluginFactory := func() (framework.Plugin, error) {
+			return pluginCopy, nil
+		}
+		if err := registry.Register("foo"+strconv.Itoa(nameIndex), filterPluginFactory); err != nil {
+			return nil, err
+		}
+		nameIndex++
+	}
```
Comment (ikaven1024) [inline pkg/scheduler/framework/runtime/framework_test.go:95]:
```suggestion
	for i, plugin := range plugins {
		pluginCopy := plugin
		filterPluginFactory := func() (framework.Plugin, error) {
			return pluginCopy, nil
		}
		if err := registry.Register("foo"+strconv.Itoa(i), filterPluginFactory); err != nil {
			return nil, err
		}
	}
```

**Comment by jwcesign** [comment]
cc @ikaven1024 , PTAL

**Review by RainbowMango — APPROVED** [review]
/approve

**Comment by RainbowMango** [comment]
We shouldn't install `mockgen` with a version suffix, because we have to maintain to the version in two places(go.mod and the script):
https://github.com/karmada-io/karmada/blob/4521a0ddd2d46dcbfdeca1d172467faf6736dcf3/hack/update-mocks.sh#L35
Just like I mentioned above: https://github.com/karmada-io/karmada/pull/2654#discussion_r1001603344

(It's too late to hold this, I'll send another PR for this.)

**Pull request merged**